File: src/gcp_calc/regex.py
import os
import re
from google.cloud import storage

os.environ["GCLOUD_PROJECT"] = "ece461-module-registry"
bucket_name = "tmr-bucket"

def find_regex_matches(regex_string):

    if (len(regex_string) == 0):
        return ""

    storage_client = storage.Client()
    blobs = storage_client.list_blobs(bucket_name)

    regex_pattern = re.compile(regex_string)

    results = []

    for blob in blobs:

        blobWritten = 0
        result_append = []
        
        if regex_pattern.match(blob.name) != None:
            moduleName = blob.name[:-4]
            result_append.append(moduleName)

            moduleVersion = (blob.metadata)['version']
            result_append.append(moduleVersion)

            blobWritten = 1
        
        if regex_pattern.match((blob.metadata)['README']) != None and blobWritten == 0:
            moduleName = blob.name[:-4]
            result_append.append(moduleName)

            moduleVersion = (blob.metadata)['version']
            result_append.append(moduleVersion)

            blobWritten = 1

        if blobWritten == 1:
            results.append(result_append)

    return results

print(find_regex_matches("require('lodash/fp/curryN')"))

File: src/gcp_calc/regex.go
package main

import (
	"context"
	"fmt"
	"regexp"

	"cloud.google.com/go/storage"
)

func Regex(regex_str string) []string {
	var matches []string
	bucketName := "tomr-bucket"

	// Create regex for string
	pattern, err := regexp.Compile(regex_str)
	if err != nil {
		fmt.Printf("Could not create regex pattern: %v\n", err)
		return matches
	}

	// Setup GCP client
	ctx := context.Background()
	client, err := storage.NewClient(ctx)
	if err != nil {
		fmt.Printf("Could not setup GCP client: %v\n", err)
		return matches
	}
	defer client.Close()

	// Create a query
	query := &storage.Query{
		Delimiter: "/",
	}

	// Get list of all objects and iterate
	blob := client.Bucket(bucketName).Objects(ctx, query)
	for {
		// Get next object and check for end of bucket
		obj, err := blob.Next()
		if err == storage.ErrObjectNotExist || err != nil{
			break
		}

		// Check match string for name, if not match check readme
		if pattern.MatchString(obj.Name) {
			fmt.Printf("Match with %s\n", obj.Name)
			matches = append(matches, obj.Name)
		} else {
			meta, err := client.Bucket(bucketName).Object(obj.Name).Attrs(ctx)
			if err != nil {
				fmt.Printf("Could not get metadata: %v\n", err)
				continue
			}
			readme, found := meta.Metadata["README"]
			if found {
				if pattern.MatchString(readme) {
					fmt.Printf("Matched in readme %s\n", obj.Name)
					matches = append(matches, obj.Name)
				}
			} else {
				fmt.Printf("readme not found")
			}
		}

	}

	// Return list of matches
	return matches

}

func main() {
	fmt.Println(Regex("(Cloudinary|lodash)"))
}
